### PR TITLE
Widgets CSS: Tweak - Each widget should load its own CSS only when needed - ED-1885

### PR DESCRIPTION
**Pull request by rotemee**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

**Comment by pacifier007** [comment]
Is this coming v3.2.0 or v3.3?

File: core/page-assets/data-managers/base.php [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+	 *
+	 * @since 3.3.0
+	 * @access protected
+	 *
+	 * @param string $asset_key
+	 *
+	 * @return void
+	 */
+	protected function save_asset_data( $asset_key ) {
+		$assets_data = get_option( self::ASSETS_DATA_KEY, [] );
```
Comment (kobizz) [inline core/page-assets/data-managers/base.php:156]: Call `get_saved_assets_data()` instead

File: core/page-assets/data-managers/base.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+		$assets_category = $this->assets_category;
+
+		if ( ! isset( $assets_data[ $content_type ] ) ) {
+			$assets_data[ $content_type ] = [];
+		}
+
+		if ( ! isset( $assets_data[ $content_type ][ $assets_category ] ) ) {
+			$assets_data[ $content_type ][ $assets_category ] = [];
+		}
+
```
Comment (kobizz) [inline core/page-assets/data-managers/base.php:167]: Do this logic inside `get_saved_assets_data()`

File: core/page-assets/data-managers/base.php [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
```
Comment (kobizz) [inline core/page-assets/data-managers/base.php:311]: Do this logic inside `get_saved_assets_data()`

File: includes/base/widget-base.php [hunk 4]
```diff
@@ -904,4 +947,58 @@ protected function deprecated_notice( $plugin_title, $since, $last = '', $replac
 		$this->end_controls_section();
 
 	}
+
+	public function register_runtime_widget( $widget_name ) {
+		if ( ! $widget_name ) {
+			return;
+		}
+
```
Comment (kobizz) [inline includes/base/widget-base.php:955]: Remove this condition

**Review by kobizz — CHANGES_REQUESTED** [review]

File: Gruntfile.js [hunk 5]
```diff
@@ -54,12 +61,16 @@ module.exports = function( grunt ) {
 	} );
 
 	grunt.registerTask( 'styles', ( isDevMode = false ) => {
+		grunt.task.run( 'create_widgets_temp_scss_files' );
```
[the remark this answers is not in the document]
Comment (rotemee) [reply Gruntfile.js:64]: Done.

File: assets/dev/scss/frontend/_widgets-global.scss [hunk 6]
```diff
@@ -1,7 +1,11 @@
-//
-// Button
-//
+// Default Hover Transition.
```
[the remark this answers is not in the document]
Comment (rotemee) [reply assets/dev/scss/frontend/_widgets-global.scss:1]: Done.

File: assets/dev/scss/frontend/frontend.scss [hunk 7]
```diff
@@ -13,6 +13,9 @@
 //Forms (Input, Textarea, Select Etc..)
 @import "forms";
 
+// Widgets Global
```
[the remark this answers is not in the document]
Comment (rotemee) [reply assets/dev/scss/frontend/frontend.scss:16]: Done.

File: core/page-assets/data-managers/base.php (hunk 1 above)
Comment (rotemee) [reply core/page-assets/data-managers/base.php:156]: Done.

File: core/page-assets/data-managers/base.php (hunk 2 above)
Comment (rotemee) [reply core/page-assets/data-managers/base.php:167]: Done.

File: core/page-assets/data-managers/base.php (hunk 3 above)
Comment (rotemee) [reply core/page-assets/data-managers/base.php:311]: Done.

File: includes/base/widget-base.php (hunk 4 above)
Comment (rotemee) [reply includes/base/widget-base.php:955]: Done.

File: core/experiments/manager.php [hunk 8]
```diff
@@ -236,6 +236,15 @@ private function add_default_features() {
 			'release_status' => self::RELEASE_STATUS_ALPHA,
 		] );
 
+		$this->add_feature( [
```
[the remark this answers is not in the document]
Comment (rotemee) [reply core/experiments/manager.php:239]: Done.

**Review by kobizz — APPROVED** [review]

**Pull request merged**